java/src/main/native/src/row_conversion.cu

```diff
@@ -304,8 +329,620 @@ copy_from_fixed_width_columns(const cudf::size_type start_row, const cudf::size_
+    auto const num_sections_y = util::div_rounding_up_unsafe(num_block_rows, 32);
+    auto const validity_data_row_length =
+        align_offset(util::div_rounding_up_unsafe(num_block_cols, 8), 8);
+    auto const total_sections = num_sections_x * num_sections_y;
+
+    int const warp_id = threadIdx.x / warp_size;
+    int const lane_id = threadIdx.x % warp_size;
+    auto const warps_per_block = std::max(1u, blockDim.x / warp_size);
+
+    // the block is divided into sections. A warp operates on a section at a time.
+    for (int my_section_idx = warp_id; my_section_idx < total_sections;
+         my_section_idx += warps_per_block) {
+      // convert to rows and cols
+      auto const section_x = my_section_idx % num_sections_x;
+      auto const section_y = my_section_idx / num_sections_x;
+      auto const relative_col = section_x * 32 + lane_id;
+      auto const relative_row = section_y * 32;
+      auto const absolute_col = relative_col + block.start_col;
+      auto const absolute_row = relative_row + block.start_row;
+      auto const cols_left = num_columns - absolute_col;
+      auto const participation_mask = __ballot_sync(0xFFFFFFFF, absolute_col < num_columns);
+
+      if (absolute_col < num_columns) {
+        auto my_data = input_nm[absolute_col] != nullptr ?
+                           input_nm[absolute_col][absolute_row / 32] :
+                           std::numeric_limits<uint32_t>::max();
+
+        // every thread that is participating in the warp has 4 bytes, but it's column-based
+        // data and we need it in row-based. So we shuffle the bits around with ballot_sync to
+        // make the bytes we actually write.
+        bitmask_type dw_mask = 1;
+        for (int i = 0; i < 32 && relative_row + i < num_rows; ++i, dw_mask <<= 1) {
+          auto validity_data = __ballot_sync(participation_mask, my_data & dw_mask);
+          // lead thread in each warp writes data
+          auto const validity_write_offset =
+              validity_data_row_length * (relative_row + i) + relative_col / 8;
+          if (threadIdx.x % warp_size == 0) {
+            if (cols_left <= 8) {
+              // write byte
+              this_shared_block[validity_write_offset] = validity_data & 0xFF;
+            } else if (cols_left <= 16) {
+              // write int16
+              *reinterpret_cast<int16_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data & 0xFFFF;
+            } else if (cols_left <= 24) {
+              // write int16 and then int8
+              *reinterpret_cast<int16_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data & 0xFFFF;
+              shared_data[validity_write_offset + 2] = (validity_data >> 16) & 0xFF;
+            } else {
+              // write int32
+              *reinterpret_cast<int32_t *>(&this_shared_block[validity_write_offset]) =
+                  validity_data;
+            }
```

Comment:
Are these if's necessary?  Normally I'd expect the output buffer to be padded to at least 4 bytes.  Maybe that's not always the case here though?

Reply:
The validity data in JCUDF is byte aligned so we can stomp other data if we don't do this. :(

Reply:
That said, the data in the shared memory doesn't have to be, so we could get around this specific spot.

Reply:
This did turn out to be no longer necessary as the data was padded out now. Updated.